File: lib/py/uapi/internal/binary/ConstructBinaryEncoding.py
from typing import List, Dict, Tuple, Set, Union
from typing import TYPE_CHECKING


from uapi.internal.binary.BinaryEncoding import BinaryEncoding
from uapi.internal.binary.CreateChecksum import create_checksum

if TYPE_CHECKING:
    from uapi.UApiSchema import UApiSchema
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration


def trace_type(type_declaration: 'UTypeDeclaration') -> list[str]:
    from uapi.internal.types.UArray import UArray
    from uapi.internal.types.UObject import UObject
    from uapi.internal.types.UStruct import UStruct
    from uapi.internal.types.UUnion import UUnion

    this_all_keys: list[str] = []

    if isinstance(type_declaration.type, UArray):
        these_keys2 = trace_type(type_declaration.type_parameters[0])
        this_all_keys.extend(these_keys2)
    elif isinstance(type_declaration.type, UObject):
        these_keys2 = trace_type(type_declaration.type_parameters[0])
        this_all_keys.extend(these_keys2)
    elif isinstance(type_declaration.type, UStruct):
        struct_fields = type_declaration.type.fields
        for struct_field_key, struct_field in struct_fields.items():
            this_all_keys.append(struct_field_key)
            more_keys = trace_type(struct_field.type_declaration)
            this_all_keys.extend(more_keys)
    elif isinstance(type_declaration.type, UUnion):
        union_tags = type_declaration.type.tags
        for tag_key, tag_value in union_tags.items():
            this_all_keys.append(tag_key)
            struct_fields = tag_value.fields
            for struct_field_key, struct_field in struct_fields.items():
                this_all_keys.append(struct_field_key)
                more_keys = trace_type(struct_field.type_declaration)
                this_all_keys.extend(more_keys)

    return this_all_keys


def construct_binary_encoding(u_api_schema: 'UApiSchema') -> 'BinaryEncoding':
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
    from uapi.internal.types.UFn import UFn

    all_keys: set[str] = set()

    functions: list[Tuple[str, UFn]] = []

    for key, value in u_api_schema.parsed.items():
        if isinstance(value, UFn):
            functions.append((key, value))

    for key, value in functions:
        all_keys.add(key)
        args = value.call.tags[key]
        for field_key, field in args.fields.items():
            all_keys.add(field_key)
            keys = trace_type(field.type_declaration)
            for key in keys:
                all_keys.add(key)

        result = value.result.tags['Ok_']
        all_keys.add('Ok_')
        for field_key, field in result.fields.items():
            all_keys.add(field_key)
            keys = trace_type(field.type_declaration)
            for key in keys:
                all_keys.add(key)

    sorted_all_keys = sorted(all_keys)

    print('Sorted all keys:')
    binary_encoding = {key: i for i, key in enumerate(sorted_all_keys)}
    final_string = "\n".join(sorted_all_keys)
    checksum = create_checksum(final_string)
    return BinaryEncoding(binary_encoding, checksum)

File: lib/java/src/main/java/uapi/internal/binary/ConstructBinaryEncoding.java
package uapi.internal.binary;

import static uapi.internal.binary.CreateChecksum.createChecksum;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

import uapi.UApiSchema;
import uapi.internal.types.UArray;
import uapi.internal.types.UFn;
import uapi.internal.types.UObject;
import uapi.internal.types.UStruct;
import uapi.internal.types.UTypeDeclaration;
import uapi.internal.types.UUnion;

public class ConstructBinaryEncoding {

    private static List<String> traceType(UTypeDeclaration typeDeclaration) {
        final var thisAllKeys = new ArrayList<String>();

        if (typeDeclaration.type instanceof UArray) {
            final var theseKeys2 = traceType(typeDeclaration.typeParameters.get(0));
            thisAllKeys.addAll(theseKeys2);
        } else if (typeDeclaration.type instanceof UObject) {
            final var theseKeys2 = traceType(typeDeclaration.typeParameters.get(0));
            thisAllKeys.addAll(theseKeys2);
        } else if (typeDeclaration.type instanceof UStruct s) {
            final var structFields = s.fields;
            for (final var entry : structFields.entrySet()) {
                final var structFieldKey = entry.getKey();
                final var structField = entry.getValue();
                thisAllKeys.add(structFieldKey);
                final var moreKeys = traceType(structField.typeDeclaration);
                thisAllKeys.addAll(moreKeys);
            }
        } else if (typeDeclaration.type instanceof UUnion u) {
            final var unionTags = u.tags;
            for (final var entry : unionTags.entrySet()) {
                final var tagKey = entry.getKey();
                final var tagValue = entry.getValue();
                thisAllKeys.add(tagKey);
                final var structFields = tagValue.fields;
                for (final var fieldEntry : structFields.entrySet()) {
                    final var structFieldKey = fieldEntry.getKey();
                    final var structField = fieldEntry.getValue();
                    thisAllKeys.add(structFieldKey);
                    final var moreKeys = traceType(structField.typeDeclaration);
                    thisAllKeys.addAll(moreKeys);
                }
            }
        }

        return thisAllKeys;
    }

    public static BinaryEncoding constructBinaryEncoding(UApiSchema uApiSchema) {
        final var allKeys = new TreeSet<String>();
        final var functions = new ArrayList<Map.Entry<String, UFn>>();

        for (final var entry : uApiSchema.parsed.entrySet()) {
            final var key = entry.getKey();
            final var value = entry.getValue();
            if (value instanceof UFn) {
                functions.add(Map.entry(key, (UFn) value));
            }
        }

        for (final var function : functions) {
            final var key = function.getKey();
            final var value = function.getValue();
            allKeys.add(key);
            final var args = value.call.tags.get(key);
            for (final var fieldEntry : args.fields.entrySet()) {
                final var fieldKey = fieldEntry.getKey();
                final var field = fieldEntry.getValue();
                allKeys.add(fieldKey);
                final var keys = traceType(field.typeDeclaration);
                keys.forEach(allKeys::add);
            }

            final var result = value.result.tags.get("Ok_");
            allKeys.add("Ok_");
            for (final var fieldEntry : result.fields.entrySet()) {
                final var fieldKey = fieldEntry.getKey();
                final var field = fieldEntry.getValue();
                allKeys.add(fieldKey);
                final var keys = traceType(field.typeDeclaration);
                keys.forEach(allKeys::add);
            }
        }

        final var sortedAllKeys = new ArrayList<>(allKeys);
        Collections.sort(sortedAllKeys);

        final var binaryEncoding = new HashMap<String, Integer>();
        for (int index = 0; index < sortedAllKeys.size(); index++) {
            final var key = sortedAllKeys.get(index);
            binaryEncoding.put(key, index);
        }

        final var finalString = String.join("\n", sortedAllKeys);
        final var checksum = createChecksum(finalString);

        return new BinaryEncoding(binaryEncoding, checksum);
    }
}